### PR TITLE
Leaner CI matrix + `test_io` deadlock

### LINKED ISSUE #2134: Prune CI matrix
Our current "CI matrix", which triggers upon PR approval, has become excessively large. With 5 Python versions (3.10–3.14) and 8 PyTorch versions (2.3.1–2.10.0), we are currently running 50 jobs per approved PR.

This leads to:
1. **Long Wait Times:** PRs take significantly longer than necessary to reach a "green" state for merging.
2. **Resource Waste:** Most of these permutations are redundant, as compatibility issues usually arise at the "bookends" (minimum vs. maximum supported versions) rather than intermediate versions.
3. **Queue Congestion:** Large matrices block other team members' jobs in the GitHub Actions queue.

### Proposed Optimization Strategy
I propose pruning the matrix as follows:

1. **Python Versions:** 
   - Retain the **oldest supported** (currently 3.10) and **latest stable** (currently 3.14) for standard testing.
3. **PyTorch Versions:**   
   - Instead of testing all 8 versions, we keep the **minimum supported** (currently 2.3.1), the standard version on JSC systems (2.9.1 on Stage 2026), and the  **latest stable** (currently 2.10.0).
3. **Scheduled Full Test:**
   - Create a separate scheduled workflow (`cron`) that runs the full matrix once a week on the `main` branch to catch edge-case regressions.

### Expected outcome
- Reduction from **50 jobs** to approximately **6 jobs** per PR.
- As a consequence, it will be possible let the matrix run on every PR commit, instead of waiting for approval to discover compatibility problems (as suggested by @brownbaerchen )

### TODO
- [ ] Refactor `.github/workflows/ci.yml` to use a sparse matrix.
- [ ] (Optional/TBD) Setup a weekly `cron` job for exhaustive compatibility testing.

### PR DESCRIPTION
**Pull request by ClaudiaComito**
## Due Diligence

- General:
    - [x]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [x] unit tests: all split configurations tested
    - [x] unit tests: multiple dtypes tested
    - [ ] **NEW** unit tests: MPS tested (1 MPI process, 1 GPU)
    - [ ] benchmarks: created for new functionality
    - [ ] benchmarks: performance improved or maintained
    - [ ] documentation updated where needed

## Description

This PR implements a leaner CI matrix as discussed in #2134. The race condition + deadlock arising in test_io with Python 3.14 has also been solved.  

Issue/s resolved: #2134 

## Changes proposed:

- refactor `ci.yaml` as per #2134 . Runs for every commit as requested by @brownbaerchen 
- keep the full matrix in `ci_full.yaml`. This will run when a PR is merged into `main`, if the PR affects the codebase
- in `test_io.test_load_zarr_group` prevent race condition by inserting a Barrier before starting to write out the zarr group. Might be related to [PEP 703](https://peps.python.org/pep-0703/) in combination with Zarr V3's asynchronous I/O. 
-

## Type of change

- Bug fix (non-breaking change which fixes an issue) + GH action refactoring

## Memory requirements

DNA
## Performance

DNA
#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

File: .github/workflows/ci_full.yaml [hunk 1]
```diff
@@ -0,0 +1,83 @@
+name: ci_full
+
+on:
+  push:
+    branches:
+      - main
```
Comment (brownbaerchen) [inline .github/workflows/ci_full.yaml]:
```suggestion
```
I prefer to know if my branch passes the CI before I do the PR. This allows for cleaner pipelining of the work. I.e. you have set the PR because you know it's ready and go do something else until you receive the review and the reviewer can get to reviewing immediately. If the CI only runs once you have started the PR, you need to come back and check and possibly work on it again, while the reviewer is confused if she can start reviewing or not.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -1,55 +1,30 @@
...
 on:
-  pull_request_review:
-    types: [submitted]
+  pull_request:
+    types: [opened, synchronize, reopened, ready_for_review]
 
 jobs:
   approved:
-    if: github.event.review.state == 'approved'
+    if: github.event.pull_request.draft == false
```
Comment (brownbaerchen) [inline .github/workflows/ci.yaml:11]: Would we not want to know if a draft PR fails the CI? Why is it a draft, maybe exactly because we know it fails. #2137 is an example of a draft PR where I want the CI to run.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -60,18 +35,22 @@ jobs:
...
         with:
           python-version: ${{ matrix.py-version }}
           architecture: x64
+
       - name: Test
         run: |
           pip install pytest
           pip install ${{ matrix.pytorch-version }} ${{ matrix.install-options }} --extra-index-url https://download.pytorch.org/whl/cpu
+          # use pytest -vv -x for debugging
           mpirun -n 3 pytest heat/
```
Comment (brownbaerchen) [inline .github/workflows/ci.yaml]:
```suggestion
          pytest heat/
```
I know we have serial tests elsewhere but I think it's more useful to have 1 and 4 ranks here than 3 and 4 ranks because this probably runs more often than the other CI.

File: .github/workflows/ci_full.yaml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        uses: actions/setup-python@83679a892e2d95755f2dac6acb0bfd1e9ac5d548 # v6.1.0
+        with:
+          python-version: ${{ matrix.py-version }}
+          architecture: x64
+
+      - name: Test
+        run: |
+          pip install pytest
+          pip install ${{ matrix.pytorch-version }} ${{ matrix.install-options }} --extra-index-url https://download.pytorch.org/whl/cpu
+          mpirun -n 3 pytest heat/
```
Comment (brownbaerchen) [inline .github/workflows/ci_full.yaml:84]:
```suggestion
          pytest heat/
```
Same comment as above

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -1,55 +1,30 @@
 name: ci
 
 on:
-  pull_request_review:
-    types: [submitted]
+  pull_request:
+    types: [opened, synchronize, reopened, ready_for_review]
```
Comment (brownbaerchen) [inline .github/workflows/ci.yaml]:
```suggestion
  push:
```
I prefer to know if my branch passes the CI before I do the PR. This allows for cleaner pipelining of the work. I.e. you have set the PR because you know it's ready and go do something else until you receive the review and the reviewer can get to reviewing immediately. If the CI only runs once you have started the PR, you need to come back and check and possibly work on it again, while the reviewer is confused if she can start reviewing or not.

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (ClaudiaComito) [inline .github/workflows/ci.yaml:11]:
```suggestion
```

**Review by brownbaerchen — DISMISSED** [review]

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -1,55 +1,30 @@
 name: ci
 
 on:
-  pull_request_review:
-    types: [submitted]
+  pull_request:
+  push:
```
Comment (ClaudiaComito) [inline .github/workflows/ci.yaml]:
```suggestion
  push:
    branches-ignore:
      - main
      - dependabot/**    
```

File: .github/workflows/ci_full.yaml (hunk 1 above)
Comment (ClaudiaComito) [inline .github/workflows/ci_full.yaml:6]:
```suggestion
      - main
      - stable      
```

**Review by brownbaerchen — DISMISSED** [review]

**Comment by brownbaerchen** [comment]
Can anybody reproduce the deadlock locally? I can't with Python 3.14 and torch==2.9.1 torchvision==0.24.1 torchaudio==2.9.1.

File: .github/workflows/ci_full.yaml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      fail-fast: false
+      matrix:
+        py-version:
+          - '3.10'
+          - '3.11'
+          - '3.12'
+          - '3.13'
+          - '3.14'
+        mpi: [ 'openmpi' ]
+        install-options: [ '.', '.[hdf5,netcdf,pandas,zarr]' ]
```
Comment (JuanPedroGHM) [inline .github/workflows/ci_full.yaml:27]: Pandas is not an optional dependency of heat

File: .github/workflows/ci_full.yaml [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+          - py-version: '3.14'
+            pytorch-version: 'torch==2.8.0 torchvision==0.23.0 torchaudio==2.8.0'
+          - py-version: '3.13'
+            pytorch-version: 'torch==2.3.1 torchvision==0.18.1 torchaudio==2.3.1'
+          - py-version: '3.13'
+            pytorch-version: 'torch==2.4.1 torchvision==0.19.1 torchaudio==2.4.1'
+          - py-version: '3.13'
+            pytorch-version: 'torch==2.5.1 torchvision==0.20.1 torchaudio==2.5.1'
+          - py-version: '3.10'
+            install-options: '.[hdf5,netcdf,pandas,zarr]'
```
Comment (JuanPedroGHM) [inline .github/workflows/ci_full.yaml:57]: Same here, please remove

File: .github/workflows/ci.yaml [hunk 9]
```diff
@@ -1,55 +1,32 @@
...
-          - py-version: '3.14'
-            pytorch-version: 'torch==2.8.0 torchvision==0.23.0 torchaudio==2.8.0'
-          - py-version: '3.13'
-            pytorch-version: 'torch==2.3.1 torchvision==0.18.1 torchaudio==2.3.1'
-          - py-version: '3.13'
-            pytorch-version: 'torch==2.4.1 torchvision==0.19.1 torchaudio==2.4.1'
-          - py-version: '3.13'
-            pytorch-version: 'torch==2.5.1 torchvision==0.20.1 torchaudio==2.5.1'
           - py-version: '3.10'
             install-options: '.[hdf5,netcdf,pandas,zarr]'
```
Comment (JuanPedroGHM) [inline .github/workflows/ci.yaml:28]: Another wild panda

File: .github/workflows/ci.yaml [hunk 10]
```diff
@@ -1,55 +1,32 @@
...
         py-version:
-          - '3.10'
-          - '3.11'
-          - '3.12'
-          - '3.13'
-          - '3.14'
+          - '3.10' # Oldest supported
+          - '3.14' # Latest stable
         mpi: [ 'openmpi' ]
         install-options: [ '.', '.[hdf5,netcdf,pandas,zarr]' ]
```
Comment (JuanPedroGHM) [inline .github/workflows/ci.yaml:21]: And another.

**Review by JuanPedroGHM — DISMISSED** [review]

**Comment by JuanPedroGHM** [comment]
> Can anybody reproduce the deadlock locally? I can't with Python 3.14 and torch==2.9.1 torchvision==0.24.1 torchaudio==2.9.1.

I could not reproduce it. Tried with torch==2.10.

**Comment by brownbaerchen** [comment]
@copilot, implement a time limit of 2 hours for the new lean CI and the full CI. Also, fix the deadlock in test_io. Do both by suggesting changes on this PR.

File: .github/workflows/ci.yaml [hunk 11]
```diff
@@ -1,55 +1,33 @@
...
-          - '3.10'
-          - '3.11'
-          - '3.12'
-          - '3.13'
-          - '3.14'
+          - '3.10' # Oldest supported
+          - '3.14' # Latest stable
         mpi: [ 'openmpi' ]
-        install-options: [ '.', '.[hdf5,netcdf,pandas,zarr]' ]
+        install-options: [ '.', '.[hdf5,netcdf,zarr]' ]
```
Comment (brownbaerchen) [inline .github/workflows/ci.yaml]:
```suggestion
        install-options: [ '.', '.[hdf5,netcdf,pandas,zarr]' ]
```
pandas added as optional dependency in #2149

File: .github/workflows/ci.yaml [hunk 12]
```diff
@@ -1,55 +1,33 @@
...
-          - 'torch==2.7.1 torchvision==0.22.1 torchaudio==2.7.1'
-          - 'torch==2.8.0 torchvision==0.23.0 torchaudio==2.8.0'
-          - 'torch==2.9.1 torchvision==0.24.1 torchaudio==2.9.1'
-          - 'torch==2.10.0 torchvision==0.25.0 torchaudio==2.10.0'
+          - 'torch==2.3.1 torchvision==0.18.1 torchaudio==2.3.1'   # Oldest supported
+          - 'torch==2.9.1 torchvision==0.24.1 torchaudio==2.9.1'   # JSC Stage 2026
+          - 'torch==2.10.0 torchvision==0.25.0 torchaudio==2.10.0' # Latest stable
         exclude:
+          - py-version: '3.10'
+            install-options: '.[hdf5,netcdf,zarr]'
```
Comment (brownbaerchen) [inline .github/workflows/ci.yaml]:
```suggestion
            install-options: '.[hdf5,netcdf,pandas,zarr]'
```

File: .github/workflows/ci_full.yaml [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+      fail-fast: false
+      matrix:
+        py-version:
+          - '3.10'
+          - '3.11'
+          - '3.12'
+          - '3.13'
+          - '3.14'
+        mpi: [ 'openmpi' ]
+        install-options: [ '.', '.[hdf5,netcdf,zarr]' ]
```
Comment (brownbaerchen) [inline .github/workflows/ci_full.yaml]:
```suggestion
        install-options: [ '.', '.[hdf5,netcdf,pandas,zarr]' ]
```

File: .github/workflows/ci_full.yaml [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+          - py-version: '3.14'
+            pytorch-version: 'torch==2.8.0 torchvision==0.23.0 torchaudio==2.8.0'
+          - py-version: '3.13'
+            pytorch-version: 'torch==2.3.1 torchvision==0.18.1 torchaudio==2.3.1'
+          - py-version: '3.13'
+            pytorch-version: 'torch==2.4.1 torchvision==0.19.1 torchaudio==2.4.1'
+          - py-version: '3.13'
+            pytorch-version: 'torch==2.5.1 torchvision==0.20.1 torchaudio==2.5.1'
+          - py-version: '3.10'
+            install-options: '.[hdf5,netcdf,zarr]'
```
Comment (brownbaerchen) [inline .github/workflows/ci_full.yaml]:
```suggestion
            install-options: '.[hdf5,netcdf,pandas,zarr]'
```

**Review by brownbaerchen — COMMENTED** [review]
Adding pandas back in as optional dependency (needs #2149)

**Review by brownbaerchen — APPROVED** [review]
